### PR TITLE
Add GetPinnableReference back to Span and ReadOnlySpan

### PR DESCRIPTION
**Pull request by ahsonkhan**
Part of https://github.com/dotnet/corefx/issues/28969

Related PR in corefx: https://github.com/dotnet/corefx/pull/29003

cc @davidfowl, @GrabYourPitchforks, @stephentoub, @KrzysztofCwalina, @pakrym, @ahsonkhan @jkotas, @VSadov, @benaadams

### CODE REVIEW

File: src/mscorlib/shared/System/ReadOnlySpan.Fast.cs [hunk 1]
```diff
@@ -266,5 +279,7 @@ public T[] ToArray()
             Buffer.Memmove(ref Unsafe.As<byte, T>(ref destination.GetRawSzArrayData()), ref _pointer.Value, (nuint)_length);
             return destination;
         }
+
+        private static T s_nullPtr = default;
```
Comment (jkotas) [inline src/mscorlib/shared/System/ReadOnlySpan.Fast.cs]: This should not be static.

File: src/mscorlib/shared/System/ReadOnlySpan.Fast.cs [hunk 2]
```diff
@@ -149,6 +149,19 @@ internal ReadOnlySpan(ref T ptr, int length)
...
 
+        /// <summary>
+        /// Returns a reference to the 0th element of the Span. If the Span is empty, returns null reference.
+        /// It can be used for pinning and is required to support the use of span within a fixed statement.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public ref readonly T GetPinnableReference()
+        {
+            if (_length == 0)
+                return ref s_nullPtr;
```
Comment (jkotas) [inline src/mscorlib/shared/System/ReadOnlySpan.Fast.cs]: This should return `Unsafe.AsRef<T>(0)` or something similar.

File: src/mscorlib/shared/System/ReadOnlySpan.Fast.cs [hunk 3]
```diff
@@ -149,6 +149,19 @@ internal ReadOnlySpan(ref T ptr, int length)
...
         }
 
+        /// <summary>
+        /// Returns a reference to the 0th element of the Span. If the Span is empty, returns null reference.
+        /// It can be used for pinning and is required to support the use of span within a fixed statement.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public ref readonly T GetPinnableReference()
+        {
+            if (_length == 0)
```
Comment (jkotas) [inline src/mscorlib/shared/System/ReadOnlySpan.Fast.cs]: It maybe better to write this as `return (_length != 0) ? ref _pointer.Value : ref Unsafe.AsRef<T>(null)`.

File: src/mscorlib/shared/System/ReadOnlySpan.Fast.cs (hunk 1 above)
Comment (VSadov) [reply src/mscorlib/shared/System/ReadOnlySpan.Fast.cs]: I wonder if it is worth caching at all - vs. computing.

Comment (jkotas) [reply src/mscorlib/shared/System/ReadOnlySpan.Fast.cs]: It is caching the wrong thing also.

**Comment by ahsonkhan** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Pull request merged**